Add HistoricalRootsBlockProof for merge till capella blocks

kdeme opened the pull request:
The way for proving these blocks was already presented/discussed a while back and in the mean while we also have test vectors added thanks to @ogenev (https://github.com/ethereum/portal-spec-tests/pull/9), so I suggest that we start thinking about a path for activating this in the network.

Aside from potentially adjusting the description in this PR and having it implemented in the clients, we will also need changes to the bridges to build these proofs and, more importantly, discuss on how we will get rid of the old headers without the proof that might already live on the network. Hence, keeping this in draft for now.

kdeme commented on history-network.md:
```diff
@@ -160,13 +160,25 @@ each receipt/transaction and re-rlp-encode it, but only if it is a legacy transa
 AccumulatorProof = Vector[Bytes32, 15]
 
-BlockHeaderProof = Union[None, AccumulatorProof]
+BeaconBlockBodyProof = Vector[Bytes32, 8] # Proof that EL block_hash in ExecutionPayload is part of BeaconBlockBody
+BeaconBlockHeaderProof = Vector[Bytes32, 3] # Proof that BeaconBlockBody root is part of BeaconBlockHeader
+HistoricalRootsProof = Vector[Bytes32, 14] # Proof that BeaconBlockHeader root is part of HistoricalRoots
+
+HistoricalRootsBlockProof = Container[
```
Not fully sure about the naming here yet, feedback requested :-). I went with the names I used also in the initial presentation.

Basically each proof is named for which structure "the leaf" is proven to be part of.  But the inverse would work also (e.g. BlockHashProof instead of BeaconBlockBodyProof).

With this current approach, we can also have a naming clash in the future when we do the same for `historical_summaries`, as there we would have then the `HistoricalSummariesProof`, which also exists in our Portal beacon chain network (https://github.com/ethereum/portal-network-specs/blob/master/beacon-chain/beacon-network.md#historicalsummaries), but it is something different there (proof that historical summaries is part of beacon state).

acolytec3 commented on history-network.md:
```diff
@@ -166,7 +166,20 @@ each receipt/transaction and re-rlp-encode it, but only if it is a legacy transa
 AccumulatorProof = Vector[Bytes32, 15]
 
-BlockHeaderProof = Union[None, AccumulatorProof]
+BeaconBlockBodyProof = Vector[Bytes32, 8] # Proof that EL block_hash in ExecutionPayload is part of BeaconBlockBody
+BeaconBlockHeaderProof = Vector[Bytes32, 3] # Proof that BeaconBlockBody root is part of BeaconBlockHeader
```
What is the reason driving 2 separate proofs here rather than just one `BeaconBlockProof` that roots the `blockhash` in the `BeaconBlock` container (since [BeaconBlock](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#beaconblock) as extended by [this] (https://github.com/ethereum/consensus-specs/blob/dev/specs/bellatrix/beacon-chain.md#beaconblockbody) contains everything you need in one proof (so would save us one 2 values in the `HistoricalRootsBlockProof` datastructure.

kdeme replied:
Yes, good point. I hadn't realized that `BeaconBlock` results in the same `hash_tree_root` as `BeaconBlockHeader`. So it looks like what you suggest will work. I will adapt.

kdeme replied:
I've updated the PR with your suggestion.

pipermerriam commented on history/history-network.md:
```diff
@@ -165,9 +165,25 @@ each receipt/transaction and re-rlp-encode it, but only if it is a legacy transa
+# Proof for EL BlockHeader after TheMerge until Capella
+HistoricalRootsBlockProof = Container[
+    beaconBlockProof: BeaconBlockProof,
+    beaconBlocRoot: Bytes32,
```
Is the the beacon block that the proof is anchored to?

kdeme replied:
As I mentioned here in this link,  https://github.com/ethereum/portal-network-specs/pull/287#discussion_r1569186217, would it be more clear to go with object naming where  the name of the proof is about what it proves (instead of "how")?

```py
BlockProofHistoricalRoots = Container[
    beaconBlockProof: BeaconBlockProofHistoricalRoots, # Proof that the BeaconBlock is part of the historical_roots and thus part of the canonical chain
    beaconBlockRoot: Bytes32,
    executionBlockProof: ExecutionBlockProof, # Proof that EL BlockHash is part of the BeaconBlock
    slot: Slot
]
```

The reason I still attach the `HistoricalRoots` part at the end, is because there is also the `HistoricalSummaries` version. This would then be:

```py
BlockProofHistoricalSummaries = Container[
    beaconBlockProof: BeaconBlockProofHistoricalSummaries, # Proof that the BeaconBlock is part of the historical_summaries and thus part of the canonical chain
    beaconBlockRoot: Bytes32,
    executionBlockProof: ExecutionBlockProof, # Proof that EL BlockHash is part of the BeaconBlock
    slot: Slot
]
```

pipermerriam reviewed (commented):
> I suspect my knowledge of the proving data structures here may not be adequate to provide fully informed feedback.

IIUC this gives us a proof format for blocks from the merge to Capella, during which the beacon chain was using `HistoricalRoots`, after which the beacon chain migrates to `HistoricalSummaries`.

I have an open question about what the `HistoricalRootsBlockProof.beaconBlockRoot` field represents.  My assumption (which may be wrong) is that the field is the beacon block for which the proof is anchored.

Assuming I'm correct about these things, I have the following suggestion.

This fork happened a long time ago, and thus all of the proofs and data are generally frozen in history.  What if we anchored *all* of these proofs to a fixed root that could be baked into every client, so that all blocks from this range would be able to be proven against a fixed root hash.  That root hash would probably be either the last block before the fork or the first block after the fork, and then all of the proofs of this type could be proven against that singular block root without need to source data from another network.

kdeme commented:
> IIUC this gives us a proof format for blocks from the merge to Capella, during which the beacon chain was using `HistoricalRoots`, after which the beacon chain migrates to `HistoricalSummaries`.

That's right.

> I have an open question about what the `HistoricalRootsBlockProof.beaconBlockRoot` field represents. My assumption (which may be wrong) is that the field is the beacon block for which the proof is anchored.

This is the root of the `BeaconBlock` of which the EL block hash is part of (`BeaconBlock` -> `BeaconBlockBody` -> `ExecutionPayload`). 
The `BeaconBlockProof` is then the SSZ merkle proof that this EL block hash is part of the `BeaconBlock` with that root. So the verify part requires this proof + the `BeaconBlock` root as root and the EL block hash as leaf.

The other part of the verification is to make sure that the BeaconBlock used is canonical. This uses the historical_roots and works pretty much the same as our accumulator pre-merge. So the main difference here is that we get to proof if a `BeaconBlock` is canonical, and not an EL block. But as an EL block basically part of the `BeaconBlock`, we can used the above explained extra step to get also the EL block (hash) proven.

The historical_roots can/will be baked in the binary as it is frozen.




> What if we anchored _all_ of these proofs to a fixed root that could be baked into every client, so that all blocks from this range would be able to be proven against a fixed root hash. That root hash would probably be either the last block before the fork or the first block after the fork, and then all of the proofs of this type could be proven against that singular block root without need to source data from another network.

I don't think I fully understand this part. Is it still applicable with above explanation?

kdeme commented on history/history-network.md:
```diff
@@ -338,5 +355,13 @@ Merkle proof with the `BlockHeader`'s block hash as leave and the
+
+The `HistoricalRootsBlockProof` is an SSZ container which holds two Merkle proofs as specified in the [SSZ Merke proofs specification](https://github.com/ethereum/consensus-specs/blob/dev/ssz/merkle-proofs.md#merkle-multiproofs).
+
+The container holds a chain of 2 proofs. This chain of proofs allows for verifying that an EL `BlockHeader` is part of the canonical chain. The only requirement is having access to the beacon chain `historical_roots`.
+The `historical_roots` is a [`BeaconState` field](https://github.com/ethereum/consensus-specs/blob/dev/specs/capella/beacon-chain.md#beaconstate) that is frozen since the Capella fork. The `HistoricalRootsBlockProof` MUST be used to verify blocks from TheMerge/Paris until the Capella fork.
```
This is to minimal to be understandable. I'll need to add some further explanation here.

pipermerriam reviewed (approved):
All of my questions are generally answered and I also apologize for asking confusing questions.  I think the biggest thing that this PR needs (which you already acquiesced to) is a more detailed explanation of the properties and structure of the proof.  (and maybe a diagram).

Lets get this merged

kdeme commented:
I have rebased the PR and updated it with renaming of the proofs + extra text clarification + some charts in 6d7a5227479bd1e8959756dbb4dcb3df3698e461,  [fbb2a43](https://github.com/ethereum/portal-network-specs/pull/287/commits/fbb2a434e44036d1981d4204c2cb48e1665c0e11) and [af07120](https://github.com/ethereum/portal-network-specs/pull/287/commits/af07120aa9485b93719ccd578827db6dd2b38139).

pipermerriam commented on history/history-network.md:
```diff
@@ -155,9 +155,25 @@ each receipt/transaction and re-rlp-encode it, but only if it is a legacy transa
+# Proof for EL BlockHeader from TheMerge until Capella
+BlockProofHistoricalRoots = Container[
+    beaconBlockProof: BeaconBlockProofHistoricalRoots, # Proof that the BeaconBlock is part of the historical_roots and thus part of the canonical chain
+    beaconBlockRoot: Bytes32,
```
Maybe add comments to these fields too which describe what they are as well so that everything has an explanation?

kdeme replied:
Added in https://github.com/ethereum/portal-network-specs/pull/287/commits/4aef56bf584992f07adbf1efa0ba45ddc5cd700d

pipermerriam reviewed (approved):
The additional explanations and the charts make a big difference for me.  Thank you.

kdeme commented:
This PR is blocked by a solution/action taken for removal on the `None` in the Union of the block proofs and also https://github.com/ethereum/portal-network-specs/issues/337

KolbyML reviewed (approved):
:shipit: looks good

kdeme commented:
I'm fine with merging this as is now as long as we later do still have a look at either:
a. Removing the `None` and thus shifting al the selector
b. or Removing the `Union` all together.

KolbyML commented:
> I'm fine with merging this as is now as long as we later do still have a look at either: a. Removing the `None` and thus shifting al the selector b. or Removing the `Union` all together.

Kim and I discussed this a bit offline. I am kind of a fan of b but we will have a wider discussion with the teams on this issue

morph-dev commented:
In general, it looks good to me.

Since there was some discussion about naming, I have an idea but I'm not sure if it was considered before.
How about naming top level proofs something like: `BlockPreMergeProof` and `BlockPreCapellaProof`? It makes it clear when each proof type should be used.
